[Batch Mode] Move supplementary output paths into a dedicated structure in FrontendInputsAndOutputs (3)

davidungar opened the pull request:
<!-- What's in this pull request? -->
In preparation for many sets of supplementary outputs, reify the set, and move it into FrontendInputsAndOutputs. Also move the computation of those.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->


<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

davidungar committed:
Tweaks to shorten performCompile* a bit more.

davidungar committed:
Move supplementary outputs to SupplementaryOutputs in FrontendInputsAndOutputs.

davidungar commented:
@swift-ci please smoke test

davidungar commented:
@swift-ci please smoke test os x platform

davidungar commented:
@swift-ci please smoke test

davidungar commented:
There’s a new diagnostic in this PR. I’ll try to write a test for it tomorrow.

jrose-apple commented on lib/Frontend/ArgsToFrontendOptionsConverter.cpp:
```diff
@@ -431,108 +432,48 @@ bool ArgsToFrontendOptionsConverter::computeFallbackModuleName() {
-    assert(outs->empty() &&
-           "cannot have main outputs for actions that don't produce outputs");
+    assert(outs->first.empty() && outs->second.empty() &&
+           "cannot have outputs for actions that don't produce outputs");
```
This isn't correct; `-typecheck` produces no main outputs but may have supplementary outputs.

davidungar replied:
Fixed, thanks for catching this.

jrose-apple commented on include/swift/AST/DiagnosticsFrontend.def:
```diff
@@ -190,6 +190,9 @@ ERROR(error_index_failed_status_check,none,
   "index output filenames do not match input source files", ())
 
+ERROR(error_too_many_supplementary_outputs,none,
+"too many supplementary outputs '%0'", (StringRef))
```
This should probably just be `too many %0 arguments` or perhaps `too many %0 arguments (expected %1, got %2)`.

davidungar replied:
Very nice! Fixed.

jrose-apple commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -202,3 +217,209 @@ OutputFilesComputer::deriveOutputFileFromParts(StringRef dir,
+  const unsigned N =
+      InputsAndOutputs.countOfFilesProducingSupplementaryOutput();
+  if (paths.size() > N) {
+    std::unique_ptr<llvm::opt::OptTable> Table = createSwiftOptTable();
```
You shouldn't have to greate a new OptTable for this; each Arg already has a reference to the full Option info in the existing table.

davidungar replied:
Thanks. Figured out a way to do it without generating a new table.

jrose-apple commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -202,3 +217,209 @@ OutputFilesComputer::deriveOutputFileFromParts(StringRef dir,
+    return None;
+  }
+  while (paths.size() < N)
+    paths.push_back(std::string());
```
This doesn't make sense; like with `-o`, there should always be exactly the right number of paths, or 0 paths.

jrose-apple replied:
(Honestly it *is* useful to be able to override previous options by having the last argument win, but that's still not what's going on here.)

davidungar replied:
Do you mean that the code should emit a diagnostic and bail out if there are paths, but not the right number?

jrose-apple replied:
I think so, yes.

davidungar replied:
OK, done, I think.

jrose-apple commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -202,3 +217,209 @@ OutputFilesComputer::deriveOutputFileFromParts(StringRef dir,
+  if (!serializedDiagnosticsPath)
+    return None;
+
+  auto objCHeaderOutputPath = determineSupplementaryOutputFilename(
```
Super nitpick: "ObjC" is one word when abbreviated as such, so its lowercase form is "objc".

davidungar replied:
Fixed.

jrose-apple commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -202,3 +217,209 @@ OutputFilesComputer::deriveOutputFileFromParts(StringRef dir,
+    return outputFilename;
+
+  // If we have a primary input, so use that as the basis for the name of the
+  // serialized diagnostics file, otherwise fall back on the
```
Comment needs updating.

davidungar replied:
Good catch, thanks! Much of the comments in that function were redundant, so I streamlined them.

jrose-apple reviewed (commented):
The next step is to sink this into the individual inputs, then?

davidungar commented:
Thanks for getting on this so quickly! You are keeping me energized.  I have a midday commitment, but otherwise will be working on these comments to get you a revision ASAP, including a test for the new diagnostic.

The next step will be the introduction of the PrimarySpecificPaths structure, which can be passed around in the compiler and holds both main and supp outputs. Then, putting that in the inputs.

davidungar commented on include/swift/Frontend/SupplementaryOutputPaths.h:
```diff
@@ -0,0 +1,77 @@
+  SupplementaryOutputPaths() = default;
+  SupplementaryOutputPaths(const SupplementaryOutputPaths &) = default;
+
+  bool empty() const {
```
With the fix below, it may not. I'll remove it if possible.

davidungar commented on include/swift/Frontend/SupplementaryOutputPaths.h:
```diff
@@ -0,0 +1,77 @@
+                           std::string referenceDependenciesFilePath,
+                           std::string serializedDiagnosticsPath,
+                           std::string loadedModuleTracePath,
+                           std::string tbdPath)
```
Yes, a problematic tradeoff. I've made the change.

davidungar commented on include/swift/Frontend/ArgsToFrontendOutputsConverter.h:
```diff
@@ -41,7 +42,8 @@ class ArgsToFrontendOutputsConverter {
         Diags(diags) {}
 
-  Optional<std::vector<std::string>> convert();
+  Optional<std::pair<std::vector<std::string>, SupplementaryOutputPaths>>
```
I love the functional style, but since this will get even more complicated when a vector of SupplementaryPaths is returned, I'm changing it as suggested.

jrose-apple replied:
I guess another alternative would be to use a one-off struct.

davidungar replied:
Yes, that would work, too. I just went with out-parameters.

davidungar commented on include/swift/Frontend/ArgsToFrontendOutputsConverter.h:
```diff
@@ -108,6 +110,50 @@ class OutputFilesComputer {
+
+  /// \return None if error
+  /// Otherwise return a vector with one entry per file producing supplementary
+  /// outputs.
```
Expanded the comment and tweaked the name. If it is still unclear, would you like to suggest something?

davidungar commented on include/swift/Frontend/ArgsToFrontendOutputsConverter.h:
```diff
@@ -108,6 +110,50 @@ class OutputFilesComputer {
+  /// Otherwise return a vector with one entry per file producing supplementary
+  /// outputs.
+  Optional<std::vector<std::string>>
+  readSupplementaryOutputArguments(swift::options::ID path_id) const;
```
Fixed, thanks.

davidungar commented on include/swift/Frontend/ArgsToFrontendOutputsConverter.h:
```diff
@@ -108,6 +110,50 @@ class OutputFilesComputer {
+      const SupplementaryOutputPaths &pathsFromFilelists,
+      const InputFile &) const;
+
+  StringRef deriveImplicitBasis(StringRef outputFilename,
```
Renamed function to: `deriveDefaultSupplementaryOutputPathExcludingExtension` and renamed `implicitBasis` to `defaultSupplementaryOutputPathExcludingExtension`. May be long, but at least it's clear.

davidungar commented:
@swift-ci please smoke test

davidungar commented:
At this point, I've taken a whack at addressing all of the suggestions. Feel free to let me know what ought to be done next. I'll also be writing a test for the new diagnostic.

davidungar commented:
@swift-ci please smoke test

davidungar commented:
@swift-ci please smoke test os x platform

davidungar commented:
@swift-ci please smoke test os x platform

davidungar commented:
@swift-ci please smoke test os x platform

davidungar commented:
@swift-ci please smoke test os x platform

davidungar commented:
@swift-ci please smoke test os x platform

davidungar commented:
@swift-ci please smoke test os x platform

davidungar commented:
@swift-ci please test

jrose-apple commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -202,3 +218,225 @@ OutputFilesComputer::deriveOutputFileFromParts(StringRef dir,
+    return None;
+  }
+  while (paths.size() < N)
+    paths.push_back(std::string());
```
…which means this can become, unconditionally, [`paths.rezise(N)`](http://en.cppreference.com/w/cpp/container/vector/resize).

davidungar replied:
Thanks for reviewing so quickly! I went down a rathole last night and had to back up.
Maybe have backed up too far. I have rewritten this routine and directly constructs a vector with N blanks if needed.

jrose-apple commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -202,3 +218,225 @@ OutputFilesComputer::deriveOutputFileFromParts(StringRef dir,
+StringRef SupplementaryOutputPathsComputer::
+    deriveDefaultSupplementaryOutputPathExcludingExtension(
+        StringRef outputFilename, const InputFile &input) const {
+  // Put the serialized diagnostics file next to the output file if possible.
```
Another specific-rather-than-generic comment.

davidungar replied:
Thank you for the catch! Fixed.

jrose-apple commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -202,3 +218,225 @@ OutputFilesComputer::deriveOutputFileFromParts(StringRef dir,
+    return pathFromArguments;
+
+  if (!Args.hasArg(emitOpt))
+    return std::string();
```
Shouldn't this be `None`?

davidungar replied:
std::string() is correct, but the code was unclear because it was using None for error. However,
there is no error condition in this version. So, I modified this function and its caller to just return a string, instead of an optional. And, I added a Doxygen comment. (The optional may come back in a future PR, but I'll comment it if so.)

jrose-apple commented on lib/Frontend/FrontendInputsAndOutputs.cpp:
```diff
@@ -284,10 +286,12 @@ void FrontendInputsAndOutputs::forEachInputProducingAMainOutputFile(
+    ArrayRef<std::string> outputFiles,
+    SupplementaryOutputPaths supplementaryOutputs) {
+  if (outputFiles.empty())
+    ;
```
Does this need to be handled explicitly at all? When is the assertion from before no longer valid?

davidungar replied:
I just pushed a new commit that cleans up these corner cases. Check it out!

jrose-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -1037,9 +1052,10 @@ static void setPrivateDiscriminatorIfNeeded(IRGenOptions &IRGenOpts,
-static bool serializeSIB(FrontendOptions &opts, SILModule *SM,
-                         ASTContext &Context, ModuleOrSourceFile MSF) {
-  const std::string &moduleOutputPath = opts.ModuleOutputPath;
+static bool serializeSIBIfNeeded(FrontendOptions &opts, SILModule *SM,
```
Looks like the wrong name snuck back into this version.

davidungar replied:
Thanks. Fixed.

davidungar commented:
@swift-ci please smoke test os x platform

davidungar commented:
@swift-ci please smoke test os x platform

davidungar committed:
Cleaner handling of cases where action produces no output or where there are no inputs.

davidungar commented:
@swift-ci please smoke test

davidungar commented:
@swift-ci please smoke test

jrose-apple commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -202,3 +224,235 @@ OutputFilesComputer::deriveOutputFileFromParts(StringRef dir,
+  Diags.diagnose(SourceLoc(), diag::error_wrong_number_of_arguments,
+                 Args.getLastArg(pathID)->getOption().getName(), N,
+                 paths.size());
+  return None;
```
Nice, I like the way this came out!

davidungar replied:
:)

jrose-apple commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -202,3 +224,235 @@ OutputFilesComputer::deriveOutputFileFromParts(StringRef dir,
+    options::ID emitOpt, std::string pathFromArguments, StringRef extension,
+    StringRef mainOutputIfUsable,
+    StringRef defaultSupplementaryOutputPathExcludingExtension) const {
+  using namespace options;
```
Nitpick: this is no longer necessary.

davidungar replied:
Thanks. Removed.

jrose-apple commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -202,3 +224,235 @@ OutputFilesComputer::deriveOutputFileFromParts(StringRef dir,
+    return std::vector<std::string>(N, std::string());
+
+  Diags.diagnose(SourceLoc(), diag::error_wrong_number_of_arguments,
+                 Args.getLastArg(pathID)->getOption().getName(), N,
```
`getPrefixedName()` produces a slightly prettier diagnostic (it says "-emit-module-path" instead of "emit-module-path").

davidungar replied:
Great! Done.

jrose-apple reviewed (approved):
Last few comments, but no need for *another* round of review after that!

davidungar commented:
Good catches! I'll make the fixes, retest, and merge. Thank you.

davidungar commented:
@swift-ci please smoke test

davidungar commented:
@swift-ci please smoke test
